cli/printlog/printlog.go

```diff
@@ -46,10 +50,244 @@ func HandlePrint(args []string) (int, error) {
 		}
 		return 0, nil
 	}
+	if *compactExecLog != "" {
+		if err := printCompactExecLog(*compactExecLog); err != nil {
+			return -1, err
+		}
+		return 0, nil
+	}
 	log.Print(usage)
 	return 1, nil
 }
 
+func printCompactExecLog(path string) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	r, err := zstd.NewReader(f)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+	slr := NewSpawnLogReconstructor(r)
+
+	for {
+		sr, err := slr.GetSpawnExec()
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("failed getting spawn exec: %s", err)
+		}
+		b, err := protojson.MarshalOptions{Multiline: true}.Marshal(sr)
+		if err != nil {
+			return fmt.Errorf("failed to marshal remote gRPC log entry: %s", err)
+		}
+		if _, err := os.Stdout.Write(append(b, []byte{'\n'}...)); err != nil {
+			return fmt.Errorf("failed to write to stdout: %s", err)
+		}
+	}
+}
+
+type SpawnLogReconstructor struct {
+	input io.Reader
+
+	hashFunc string
+	files    map[int32]*spb.File
+	dirs     map[int32]*reconstructedDir
+	symlinks map[int32]*spb.File
+	sets     map[int32]*spb.ExecLogEntry_InputSet
+}
+
+type reconstructedDir struct {
+	path  string
+	files []*spb.File
+}
+
+func NewSpawnLogReconstructor(input io.Reader) *SpawnLogReconstructor {
+	return &SpawnLogReconstructor{
+		input:    input,
+		hashFunc: "",
+		files:    make(map[int32]*spb.File),
+		dirs:     make(map[int32]*reconstructedDir),
+		symlinks: make(map[int32]*spb.File),
+		sets:     make(map[int32]*spb.ExecLogEntry_InputSet),
+	}
+}
+
+func (slr *SpawnLogReconstructor) GetSpawnExec() (*spb.SpawnExec, error) {
+	entry := &spb.ExecLogEntry{}
+	br := bufio.NewReader(slr.input)
```

Comment:
since this buffers `slr.input` I think we might need to store `br` as a struct member and reuse the same reader for each function call

Reply:
ah I missed that. Good call!